junit/src/main/java/cucumber/runtime/junit/JUnitReporter.java

```diff
@@ -141,6 +152,84 @@ private void addFailure(Result result) {
         }
         failedStep = true;
         stepNotifier.addFailure(error);
-        executionUnitNotifier.addFailure(error);
+        pickleRunnerNotifier.addFailure(error);
+    }
+
+    private interface TestNotifier {
+
+        void fireTestStarted();
+
+        void addFailure(Throwable error);
+
+        void fireTestIgnored();
+
+        void fireTestFinished();
+    }
+
+
+    private static final class NoTestNotifier implements TestNotifier {
+
+        @Override
+        public void fireTestStarted() {
+            // Does nothing
+        }
+
+        @Override
+        public void addFailure(Throwable error) {
+            // Does nothing
+        }
+
+        @Override
+        public void fireTestIgnored() {
+            // Does nothing
+        }
+
+        @Override
+        public void fireTestFinished() {
+            // Does nothing
+        }
+    }
+
+    static class EachTestNotifier implements TestNotifier {
+        private final RunNotifier notifier;
+
+        private final Description description;
+
+        EachTestNotifier(RunNotifier notifier, Description description) {
+            this.notifier = notifier;
+            this.description = description;
+        }
+
+        public void addFailure(Throwable targetException) {
+            if (targetException instanceof MultipleFailureException) {
+                addMultipleFailureException((MultipleFailureException) targetException);
+            } else if (targetException instanceof AssumptionViolatedException) {
```

Comment:
The different versions of `AssumtionViolatedException`s are kept in the array [`PENDING_EXCEPTIONS`](https://github.com/cucumber/cucumber-jvm/blob/master/core/src/main/java/cucumber/runtime/Runtime.java#L36-L38), [`isPending()`](https://github.com/cucumber/cucumber-jvm/blob/master/core/src/main/java/cucumber/runtime/Runtime.java#L239) searches the array, it could done here to (if also removing the cast of the argument to `addFailedAssumption` - as it uses its argument it could accept a `Throwable`)

As the [`PendingException`](https://github.com/cucumber/cucumber-jvm/blob/master/core/src/main/java/cucumber/api/PendingException.java#L4-L5) now is identified with an annotation instead of as a member of `PENDING_EXCEPTIONS`, there is a case for renaming `PENDING_EXCEPTIONS` to something with `ASSUMPTION_VIOLATED` or something - but it does not directly has anything to do with this pull request.